python/mxnet/optimizer.py

```diff
@@ -342,17 +342,20 @@ def create_state(self, index, weight):
         momentum = None
         weight_master_copy = None
         if self.multi_precision and weight.dtype == numpy.float16:
+            assert(weight.stype == 'default'), \
```

Comment:
Basically we want to add sparse update support for optimzers one by one. It either requires new kernel for the operator or requires some changes in the python optimizer definition. For now if an old optimizer receives a sparse input, it's going to convert input to dense and executes in the old way. I'll update the document of optimizer to indicate whether it supports sparse update so that people know what to expect.